### ISSUE TITLE
Refactor Argument Clinic support code into libclinic

### ISSUE THREAD

**Issue by vstinner** [python/cpython#113299]
Argument Clinic was added 10 years ago. The script `clinic.py` grown from around 2500 lines to around 6700 lines.

I propose to split this large file into sub-files to ease the maintenance of Argument Clinic.

### Linked PRs
* gh-113300
* gh-113309
* gh-113526
* gh-115542

**Comment by erlend-aasland** [comment]
This has also been discussed before. See https://github.com/python/cpython/issues/104683#issuecomment-1633273038.

**Comment by erlend-aasland** [comment]
I think it is a really good idea to split `clinic.py` into multiple files. The C preprocessor stuff is already separated out (`cpp.py`). We should do this in iterations; not everything in one big PR. Perhaps it is easiest to start with the converters; it's a large chunk of `clinic.py`, and I think it would be pretty easy to separate it from the rest. Separating out the application is also a good start.

**Comment by AlexWaygood** [comment]
+1 to splitting clinic.py up, and +1 to doing it incrementally rather than in one big PR.

Previously I've held off from doing something like this because clinic.py's use of global variables means that lots of parts of the file are tightly coupled to other parts in surprising and unfortunate ways. In #113300, `warn()` and `fail()` still use a global `clinic` variable in `utils.py`, but that variable is now altered from code in other files, making it even harder to understand what's going on. That feels worse, and it feels like we're avoiding solving the actual problem here, which is that clinic.py just shouldn't really be using global variables like this in the first place.

_But_ maybe it's acceptable for now, if splitting the code up into several files is a higher priority :) Getting rid of the global variables is hard!

**Comment by erlend-aasland** [comment]
IMO, getting rid of the global variables should be a prerequisite to splitting things up. We don't need to rush this; clinic.py has been a single file since it started out in life. Argument Clinic is already hard to comprehend, so **any refactoring, including splitting up** should make it _more readable_, not the other way around. Readability is IMO the highest priority.

**Comment by vstinner** [comment]
@erlend-aasland:
> Moreover, doing this work in chunks/iterations, instead of everything in one go, will minimise the impact on ongoing Argument Clinic work.

Ok, I created PR #113309 to create libclinic.

@erlend-aasland:
> Perhaps it is easiest to start with the converters; it's a large chunk of clinic.py, and I think it would be pretty easy to separate it from the rest. Separating out the application is also a good start.

It's quite complicated to move code with most dependencies first. It's simpler to start to move which has no dependency and move it out of clinic.py. That's how I created PR #113300.

New PR #113309 creates `utils.py`: ClinicError, fail() and warn() which have no dependency.

@AlexWaygood:
> But maybe it's acceptable for now, if splitting the code up into several files is a higher priority :) Getting rid of the global variables is hard!

Such "split clinic.py" refactoring work is already very complicated. I would prefer to not put more constraints or "pre-requirements".

@erlend-aasland:
> We don't need to rush this; clinic.py has been a single file since it started out in life.

On my side, I'm not really volunteer to work on clinic.py before it becomes a little more easier to be maintained. For me, it means to split the single file into many smaller files.

What I see is that "Argument Clinic should be refactoring" is repeated over the years, but so far, nobody did it since it's a big piece of work, and more constraints were added each time ("that should be done first").

I was surprised, test_clinic is quite big. Moreover, AC is now also tested by mypy. Nowadays, the risk of introducing regression by moving code is lower.

I'm not comfortable that "refactoring" became a requirement to implement new features: https://github.com/python/cpython/pull/113160#issuecomment-1859726036 and that "refactoring" now has requirements on "more refactoring".

**Comment by AlexWaygood** [comment]
> What I see is that "Argument Clinic should be refactoring" is repeated over the years, but so far, nobody did it since it's a big piece of work, and more constraints were added each time ("that should be done first").

I don't think this is really fair or accurate. Erlend and I have done quite a lot of incremental refactoring over the past year, and the AC code now looks pretty different as a result. It's true that we haven't changed much about the overall structure of the code yet, but you can't blame us for doing the easier things first :)

**Comment by erlend-aasland** [comment]
> What I see is that "Argument Clinic should be refactoring" is repeated over the years, but so far, nobody did it since it's a big piece of work [...]

That is not true; we've done a lot of _small_ refactorings over the last 1,5 year. We deliberately did not do it in one big PR. Several helper functions have been refactored out. We started with the largest and most complex functions. However, a lot of work still remains. That does not mean nothing has been done; that's just a false statement.

**Comment by erlend-aasland** [comment]
> I'm not comfortable that "refactoring" became a requirement to implement new features: https://github.com/python/cpython/pull/113160#issuecomment-1859726036 and that "refactoring" now has requirements on "more refactoring".

This is also a false statement; nobody said refactoring is a hard requirement. Please read the relevant comment again. Please stick to the facts.

**Comment by vstinner** [comment]
@erlend-aasland:
> IMO, getting rid of the global variables should be a prerequisite to splitting things up.

Ok. In that case, I prefer to close this issue for now, until someone has time to fix address the issue of global variables.

**Comment by erlend-aasland** [comment]
Since last week, we've done a lot of yak-shaving (and there is more to be done). A bare bones _libclinic_ has been established, containing most of the stateless formatting code. The next steps will be to move `ClinicError` into `libclinic/errors.py`, and rename `cpp.py` to `libclinic/cpp.py`.

**Comment by erlend-aasland** [comment]
`libclinic/errors.py` was established via #113525.

**Comment by erlend-aasland** [comment]
Superseded by #113317
